File: web/src/types/apikey.ts
export interface APIKey {
  id: string
  user_id: string
  workspace_id: string
  name: string
  key_prefix: string
  scopes: string[]
  last_used_at?: string | null
  request_count: number
  rate_limit?: number | null
  expires_at?: string | null
  created_at: string
}

export interface CreateAPIKeyRequest {
  name: string
  scopes: string[]
  expires_at?: string
}

export interface CreateAPIKeyResponse {
  api_key: APIKey
  key: string
}

export const API_KEY_SCOPES = [
  { value: "links:read", label: "Links (Read)", description: "View links" },
  { value: "links:write", label: "Links (Write)", description: "Create, update, delete links" },
  { value: "domains:read", label: "Domains (Read)", description: "View custom domains" },
  { value: "domains:write", label: "Domains (Write)", description: "Manage custom domains" },
  { value: "analytics:read", label: "Analytics (Read)", description: "View analytics data" },
  { value: "bio_pages:read", label: "Bio Pages (Read)", description: "View bio pages" },
  { value: "bio_pages:write", label: "Bio Pages (Write)", description: "Manage bio pages" },
  { value: "qr:read", label: "QR Codes (Read)", description: "View QR codes" },
  { value: "qr:write", label: "QR Codes (Write)", description: "Generate QR codes" },
  { value: "webhooks:read", label: "Webhooks (Read)", description: "View webhooks" },
  { value: "webhooks:write", label: "Webhooks (Write)", description: "Manage webhooks" },
] as const

File: internal/models/api_key.go
package models

import (
	"time"

	"github.com/google/uuid"
	"github.com/link-rift/link-rift/internal/repository/sqlc"
)

// ValidAPIKeyScopes defines all valid API key scopes.
var ValidAPIKeyScopes = []string{
	"links:read",
	"links:write",
	"domains:read",
	"domains:write",
	"analytics:read",
	"bio_pages:read",
	"bio_pages:write",
	"qr:read",
	"qr:write",
	"webhooks:read",
	"webhooks:write",
}

type APIKey struct {
	ID           uuid.UUID  `json:"id"`
	UserID       uuid.UUID  `json:"user_id"`
	WorkspaceID  uuid.UUID  `json:"workspace_id"`
	Name         string     `json:"name"`
	KeyHash      string     `json:"-"`
	KeyPrefix    string     `json:"key_prefix"`
	Scopes       []string   `json:"scopes"`
	LastUsedAt   *time.Time `json:"last_used_at,omitempty"`
	RequestCount int64      `json:"request_count"`
	RateLimit    *int32     `json:"rate_limit,omitempty"`
	ExpiresAt    *time.Time `json:"expires_at,omitempty"`
	CreatedAt    time.Time  `json:"created_at"`
}

type CreateAPIKeyInput struct {
	Name      string   `json:"name" binding:"required,min=1,max=100"`
	Scopes    []string `json:"scopes" binding:"required,min=1"`
	ExpiresAt *string  `json:"expires_at,omitempty"`
}

type CreateAPIKeyResponse struct {
	APIKey *APIKey `json:"api_key"`
	Key    string  `json:"key"`
}

func APIKeyFromSqlc(k sqlc.ApiKey) *APIKey {
	ak := &APIKey{
		ID:           k.ID,
		UserID:       k.UserID,
		Name:         k.Name,
		KeyHash:      k.KeyHash,
		KeyPrefix:    k.KeyPrefix,
		Scopes:       k.Scopes,
		RequestCount: k.RequestCount,
	}
	if k.WorkspaceID.Valid {
		ak.WorkspaceID = uuid.UUID(k.WorkspaceID.Bytes)
	}
	if k.LastUsedAt.Valid {
		t := k.LastUsedAt.Time
		ak.LastUsedAt = &t
	}
	if k.RateLimit.Valid {
		v := k.RateLimit.Int32
		ak.RateLimit = &v
	}
	if k.ExpiresAt.Valid {
		t := k.ExpiresAt.Time
		ak.ExpiresAt = &t
	}
	if k.CreatedAt.Valid {
		ak.CreatedAt = k.CreatedAt.Time
	}
	return ak
}

func (k *APIKey) IsExpired() bool {
	if k.ExpiresAt == nil {
		return false
	}
	return time.Now().After(*k.ExpiresAt)
}

func (k *APIKey) HasScope(scope string) bool {
	for _, s := range k.Scopes {
		if s == scope {
			return true
		}
	}
	return false
}

func IsValidScope(scope string) bool {
	for _, s := range ValidAPIKeyScopes {
		if s == scope {
			return true
		}
	}
	return false
}
